data-structures/hashmap.md

```diff
@@ -295,27 +297,66 @@ The Filecoin HAMT _does not_ use an explicit root block (`HashMapRoot`) to encod
 * `bitWidth`: The Filecoin HAMT fixes the bit width to `5`, meaning that each node of the HAMT can contain up to `2`<sup>`5`</sup> (`32`) elements containing either buckets or links to child nodes.
 * `bucketSize`: The Filecoin HAMT fixes the maximum length of its buckets to `3`, meaning a maximally full HAMT leaf node can contain `32 x 3` (`96`) key/value pairs.
 
+### Map / Bitfield layout
+
+Rather than encoding the HAMT node's `map` field as the full sequence of byte-to-element mapping with little-endian byte addressing, the Filecoin HAMT uses the byte layout of a Go [`big.Int`](https://golang.org/pkg/math/big/#Int) as a bitfield to represent the HAMT node's `map`. In practice this means that the byte array length can vary depending on which bits are set within the bitfield. Implementations must use the same encoding and decoding rules as `big.Int` to handle this field. See [this discussion thread](https://github.com/filecoin-project/go-hamt-ipld/issues/54#issuecomment-670314664) for an exploration of the specifics.
+
 ### Block layout
 
-An IPLD schema representing the Filecoin HAMT varies from the IPLD HashMap [schema](#Schema) so any implementation needing to read Filecoin HAMT blocks will need to handle its specific layout:
+The IPLD schema describing the Filecoin HAMT varies from the IPLD HashMap [schema](#Schema) and also between version of the Filecoin Actors, so any implementation needing to read Filecoin HAMT blocks will need to handle its specific layout (both layouts if historical data is important to create / read).
+
+The **v3** form of the Filecoin HAMT has the same IPLD layout as the IPLD HashMap except for the lack of an explicit root node to describe the parameters. A Filecoin HAMT's root is `HashMapNode`, which is the same as all other nodes and parameters are implicit.
+
+Note that there is currently no limitation on the types available for storage as `value`s as long as they can be decoded from the bytes. In practice, the Filecoin HAMT is used to store inline objects rather than links to objects.
+
+#### Actors v0.9 and v2 form
+
+Prior to **v3** the `Element` union was keyed with `"0"` and `"1"`.
```

Comment:
Did you consider keeping v2 as just this "keyed" note relative to v3? Because the schema is otherwise identical, so it's a bit repetitive. Assuming most people will be interested in v3, I'd lean towards keeping the spec doc shorter. The flip side is people wanting to copy-paste the v2 schema, so I agree with keeping this as-is if you were concerned about copy-pasting.

Reply:
0h, yeah, it's kind of verbose. And another problem is repeated symbols so if the .md is validated (both our schema implementations can extract schema from markdown docs) then you get overlapping symbols.
What I did in the Filecoin schema docs, inherited by @willscott's naming in statediff is versioned the symbols and I could do that here and only have the overlapping entries unversioned.  So we could have `FilecoinHAMTNodeV0` and `FilecoinHAMTNodeV3` as the different roots and they'd have `V0` and `V3` `Element` types but share `Bucket` and `Element` types. Gets kind of ugly but could be expressed without duplication and be properly unique throughout the doc.
@warpfork what think you in terms of expressiveness?